compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/descriptors/JvmDeclarationFactory.kt

```diff
@@ -152,6 +155,37 @@ class JvmDeclarationFactory(
         else
             getFieldForObjectInstance(singleton)
 
+    private fun IrClass.allFieldsAreJvmField() =
+        declarations.all { it !is IrProperty || it.backingField?.hasAnnotation(JvmAbi.JVM_FIELD_ANNOTATION_FQ_NAME) == true }
+
+    // We don't move fields to interfaces unless all fields are annotated with @JvmField.
+    // It is an error to annotate only some of the fields of an interface companion with @JvmField.
+    fun getStaticBackingFieldParent(irClass: IrClass): IrClass =
+        if (irClass.isCompanion && (!irClass.parentAsClass.isJvmInterface || irClass.allFieldsAreJvmField()))
```

Comment:
I wonder if we can get rid of the linear scan needed in the `allFieldsAreJvmField` criterion, precisely because either all or none of the properties are `@JvmField`. I.e. only look at the JvmField-ness of the property whose field we're moving, not at all of them.

Reply:
Done.